Source/WebCore/page/Navigation.cpp

```diff
@@ -122,14 +122,26 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
     RefPtr currentItem = frame()->history().currentItem();
     if (!currentItem)
         return;
-    if (previousWindow) {
+    // For main frames we can still rely on the page b/f list. However for subframes we need below logic to not lose the bookkeeping done in the previous window.
+    if (previousWindow && !frame()->isMainFrame()) {
         Ref previousNavigation = previousWindow->protectedNavigation();
-        if (previousNavigation->m_entries.size() && navigationType && navigationType == NavigationNavigationType::Reload && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
+        bool shouldProcessPreviousNavigationEntries = [&]() {
+            if (!previousNavigation->m_entries.size())
+                return false;
+            if (!(navigationType == NavigationNavigationType::Reload || navigationType == NavigationNavigationType::Push))
```

Comment:
if (a != b && a != c) rather than if (!(a == b || a == c)) would be easier to read, I think.

Reply:
Done.